### PR TITLE
feat: 🎸 Add enableStat to transferRestriction namespace

### PR DESCRIPTION
**Pull request by polymath-eric**
### Description

Adds in `enableStat` method to count and percentage transfer restriction namespace.

### JIRA Link

[DA-326](https://polymath.atlassian.net/browse/DA-326)

### Checklist

- [ ] Updated the Readme.md (if required) ?

### CODE REVIEW

File: src/api/entities/Asset/TransferRestrictions/Percentage.ts [hunk 1]
```diff
@@ -46,6 +47,11 @@ export class Percentage extends TransferRestrictionBase<TransferRestrictionType.
    */
   public declare removeRestrictions: NoArgsProcedureMethod<BigNumber>;
 
+  /**
+   * Enables investor ownership percentage statistic for the Asset, which is required before creating restrictions
+   */
+  public declare enableStat: ProcedureMethod<Omit<AddPercentStatParams, 'type'>, void>;
```
Comment (polymath-eric) [inline src/api/entities/Asset/TransferRestrictions/Percentage.ts:53]: Ideally this would be a NoArgsProcedure, but I haven't figured out how to make that work when Count needs an arg.

Right now this requires the user invoke the function with an empty object, like:
`asset.transferRestrictions.percentage.enableStat({});`

I based this off TransferRestriction, but in that example they both need args.

Comment (polymath-eric) [reply src/api/entities/Asset/TransferRestrictions/Percentage.ts:57]: Actually seems like a NoArgsProcedure is working after all. Not sure what was different when I tried it yesterday.

File: src/api/procedures/__tests__/removeAssetStat.ts [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  });
+
+  it('should add an setAssetStats transaction to the queue', async () => {
+    args = {
+      type: StatType.Balance,
+      ticker,
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const currentStats = { toArray: () => [], indexOf: () => 0 } as any;
```
Comment (polymath-eric) [inline src/api/procedures/__tests__/removeAssetStat.ts:129]: Looking at making this into a proper mock

File: src/api/procedures/addTransferRestriction.ts [hunk 3]
```diff
@@ -250,10 +203,15 @@ export async function prepareStorage(
     return cmpStat === type;
```
Comment (prashantasdeveloper) [inline src/api/procedures/addTransferRestriction.ts:203]: Same here. We need to check the claimType and issuer as well

File: src/api/procedures/removeAssetStat.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+      (type === StatType.Count && r.isMaxInvestorCount) ||
+      (type === StatType.Balance && r.isMaxInvestorOwnership)
+    ) {
+      throw new PolymeshError({
+        code: ErrorCode.UnmetPrerequisite,
+        message:
+          'This statistics cannot be removed because a TransferRequirement is currently using it',
+      });
+    }
+  });
```
Comment (prashantasdeveloper) [inline src/api/procedures/removeAssetStat.ts:133]: For `ClaimCount` and `ClaimOwnership` compliance types, you need to consider the claimTypes as well. Also for `MaxInvestorCount` and `MaxInvestorOwnership` you need to check if claimType is `null`

**Review by prashantasdeveloper — CHANGES_REQUESTED** [review]

File: src/utils/internal.ts [hunk 5]
```diff
@@ -1174,3 +1197,237 @@ export function assertTickerValid(ticker: string): void {
...
+    issuer: { did: issuerDid },
+    claimType,
+  } = claimIssuer;
+  if (type === StatType.ScopedCount && transferCondition.isClaimCount) {
+    const [rawClaim, issuer] = transferCondition.asClaimCount;
+    const restrictionIssuerDid = identityIdToString(issuer);
+    const claim = statsClaimToStatClaimUserType(rawClaim);
+    if (restrictionIssuerDid === issuerDid && claim.type === claimType) {
+      return true;
+    }
```
Comment (monitz87) [inline src/utils/internal.ts:1280]: These three things are repeated in both blocks

File: src/utils/internal.ts [hunk 6]
```diff
@@ -1174,3 +1197,237 @@ export function assertTickerValid(ticker: string): void {
...
+  }
+  return a.eq(b);
+}
+
+/**
+ * @hidden
+ */
+export function compareTransferRestrictionToInput(
+  transferRestriction: PolymeshPrimitivesTransferComplianceTransferCondition,
+  value: BigNumber | ClaimCountRestrictionValue,
```
Comment (monitz87) [inline src/utils/internal.ts:1326]: Shouldn't this also be `ClaimPercentageRestrictionValue`?

File: src/utils/internal.ts [hunk 7]
```diff
@@ -1174,3 +1197,237 @@ export function assertTickerValid(ticker: string): void {
...
+    type === TransferRestrictionType.Percentage
+  ) {
+    const currentOwnership = permillToBigNumber(transferRestriction.asMaxInvestorOwnership);
+    return currentOwnership.eq(value as BigNumber);
+  } else if (transferRestriction.isClaimCount && type === TransferRestrictionType.ClaimCount) {
+    const [statClaim, rawIssuerId, rawMin, maybeMax] = transferRestriction.asClaimCount;
+    const issuerDid = identityIdToString(rawIssuerId);
+    const min = u64ToBigNumber(rawMin);
+    const max = maybeMax.isSome ? u64ToBigNumber(maybeMax.unwrap()) : undefined;
+    const castedValue = value as ClaimCountRestrictionValue;
```
Comment (monitz87) [inline src/utils/internal.ts:1343]: Why are these assertions necessary?

File: src/utils/internal.ts [hunk 8]
```diff
@@ -1174,3 +1197,237 @@ export function assertTickerValid(ticker: string): void {
...
+      castedValue.min.eq(min) &&
+      compareOptionalBigNumbers(max, castedValue.max) &&
+      castedValue.claim.type === getClaimType(statClaim) &&
+      issuerDid === castedValue.issuer.did
+    );
+  } else if (
+    transferRestriction.isClaimOwnership &&
+    type === TransferRestrictionType.ClaimPercentage
+  ) {
+    const castedValue = value as ClaimCountRestrictionValue;
```
Comment (monitz87) [inline src/utils/internal.ts:1355]: This should be `ClaimPercentageRestrictionValue` (assuming the casts are necessary)

File: src/utils/internal.ts [hunk 9]
```diff
@@ -1174,3 +1197,237 @@ export function assertTickerValid(ticker: string): void {
...
+    const [statClaim, rawIssuerId, rawMin, rawMax] = transferRestriction.asClaimOwnership;
+    const issuerDid = identityIdToString(rawIssuerId);
+
+    const min = permillToBigNumber(rawMin);
+    const max = permillToBigNumber(rawMax);
+
+    return !!(
+      castedValue.min.eq(min) &&
+      compareOptionalBigNumbers(castedValue.max, max) &&
```
Comment (monitz87) [inline src/utils/internal.ts:1364]: These aren't optional

File: src/utils/internal.ts [hunk 10]
```diff
@@ -1174,3 +1197,237 @@ export function assertTickerValid(ticker: string): void {
...
+      castedValue.claim.type === getClaimType(statClaim) &&
+      issuerDid === castedValue.issuer.did
+    );
+  }
```
Comment (monitz87) [inline src/utils/internal.ts:1368]: There's a lot of repeated code between the last two blocks, you should be able to reuse some of it, or at least destructure `castedValue` to avoid repeating it so much

File: src/utils/internal.ts [hunk 11]
```diff
@@ -1174,3 +1197,237 @@ export function assertTickerValid(ticker: string): void {
...
+    const currentOwnership = permillToBigNumber(transferRestriction.asMaxInvestorOwnership);
+    return currentOwnership.eq(value as BigNumber);
+  } else if (transferRestriction.isClaimCount && type === TransferRestrictionType.ClaimCount) {
+    const [statClaim, rawIssuerId, rawMin, maybeMax] = transferRestriction.asClaimCount;
+    const issuerDid = identityIdToString(rawIssuerId);
+    const min = u64ToBigNumber(rawMin);
+    const max = maybeMax.isSome ? u64ToBigNumber(maybeMax.unwrap()) : undefined;
+    const castedValue = value as ClaimCountRestrictionValue;
+
+    return !!(
```
Comment (monitz87) [inline src/utils/internal.ts:1345]: The !! shouldn't be necessary

File: src/utils/internal.ts [hunk 12]
```diff
@@ -1174,3 +1197,237 @@ export function assertTickerValid(ticker: string): void {
...
+    return transferRestrictionType === TransferRestrictionType.Percentage;
+  } else if (opType === StatisticsOpType.ClaimCount) {
+    return transferRestrictionType === TransferRestrictionType.ClaimCount;
+  } else {
+    return transferRestrictionType === TransferRestrictionType.ClaimPercentage;
+  }
+}
+
+/**
+ *@ hidden
```
Comment (monitz87) [inline src/utils/internal.ts:1393]:
```suggestion
 * @hidden
```

File: src/utils/internal.ts [hunk 13]
```diff
@@ -1174,3 +1197,237 @@ export function assertTickerValid(ticker: string): void {
...
+  const { type, claimIssuer } = args;
+
+  let rawOp;
+  if (type === TransferRestrictionType.Count || type === TransferRestrictionType.ClaimCount) {
+    rawOp = statisticsOpTypeToStatOpType(StatisticsOpType.Count, context);
+  } else {
+    rawOp = statisticsOpTypeToStatOpType(StatisticsOpType.Balance, context);
+  }
+
+  const rawIssuer = claimIssuer ? claimIssuerToMeshClaimIssuer(claimIssuer, context) : undefined;
```
Comment (monitz87) [inline src/utils/internal.ts:1412]: Can't you use optionize here?

Comment (monitz87) [reply src/utils/internal.ts:1412]: (I'm assuming `claimIssuer` is an `Option`)

File: src/utils/internal.ts [hunk 14]
```diff
@@ -1174,3 +1197,237 @@ export function assertTickerValid(ticker: string): void {
...
+  currentStats: BTreeSet<PolymeshPrimitivesStatisticsStatType>,
+  neededStat: PolymeshPrimitivesStatisticsStatType
+): void {
+  const needStat = ![...currentStats].find(s => neededStat.eq(s));
+
+  if (needStat) {
+    throw new PolymeshError({
+      code: ErrorCode.UnmetPrerequisite,
+      message:
+        'The appropriate stat type for this restriction is not set for the Asset. Try calling enableStat in the namespace first',
```
Comment (monitz87) [inline src/utils/internal.ts]:
```suggestion
        "The stat type required for this restriction is not enabled for the Asset. Try calling `enableStat` in the restriction's namespace first",
```

**Review by monitz87 — CHANGES_REQUESTED** [review]
I would open the discussion on whether to rename `enableStat` to just `enable`. We're trying to hide the whole statistics thing from the user

File: src/utils/conversion.ts [hunk 15]
```diff
@@ -3771,16 +3771,72 @@ export function createStat2ndKey(
...
+      if (!condition.asClaimCount[0].isJurisdiction) {
+        return undefined;
+      } else {
+        return condition.asClaimCount[0].asJurisdiction;
+      }
+    } else if (condition.isClaimOwnership) {
+      if (!condition.asClaimOwnership[0].isJurisdiction) {
+        return undefined;
+      }
+      return condition.asClaimOwnership[0].asJurisdiction;
```
Comment (monitz87) [inline src/utils/conversion.ts:3796]: Use destructuring here

File: src/utils/conversion.ts [hunk 16]
```diff
@@ -3771,16 +3771,72 @@ export function createStat2ndKey(
...
+      return -1;
+    }
+    if (bClaim.isNone) {
+      return 1;
+    }
+    const aCode = meshCountryCodeToCountryCode(aClaim.unwrap());
+    const bCode = meshCountryCodeToCountryCode(bClaim.unwrap());
+
+    const countryOrder = Object.values(CountryCode);
+    return countryOrder.indexOf(aCode) - countryOrder.indexOf(bCode);
```
Comment (monitz87) [inline src/utils/conversion.ts:3825]: You can just use `aCode.localeCompare(bCode)`

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/entities/Asset/TransferRestrictions/Percentage.ts [hunk 17]
```diff
@@ -20,9 +21,12 @@ export class Percentage extends TransferRestrictionBase<TransferRestrictionType.
   protected type = TransferRestrictionType.Percentage as const;
 
   /**
-   * Add a Percentage Transfer Restriction to this Asset
+   * Add a Percentage Transfer Restriction to this Asset. This limits the total percentage of the float
```
[the remark this answers is not in the document]
Comment (polymath-eric) [reply src/api/entities/Asset/TransferRestrictions/Percentage.ts:23]: https://www.investopedia.com/ask/answers/what-is-companys-float/

Lets us supply though. I think I misused the term also

File: src/api/entities/Asset/TransferRestrictions/ClaimCount.ts [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+   */
+  public declare removeRestrictions: NoArgsProcedureMethod<BigNumber>;
+
+  /**
+   * Enables investor count statistic for the Asset scope by a claim, which is required before creating restrictions
+   *
+   * The number of investors need to be passed in, as the chain will not calculate the initial value for that stat,
+   * it only increments / decrements the counters as transactions happen.
+   * For `Affiliate` and `Accredited` scoped stats the number of investor who are and are not
+   * need to be given. For `Jurisdiction` the amount of holders for each CountryCode need to be given.
```
[the remark this answers is not in the document]
Comment (polymath-eric) [reply src/api/entities/Asset/TransferRestrictions/ClaimCount.ts]: It needs the amount that are, as well as the amount that aren't. So accredited + non accredited.

Reworded it a bit, hopefully thats better

File: src/utils/internal.ts (hunk 13 above)
Comment (polymath-eric) [reply src/utils/internal.ts:1411]: It is, but its actually Option<ITuple>, which made it trickier than dealing with just a normal tuple and letting polkadot.js deal with it.

We could look into making an ITuple conversion util (or maybe I am missing it), but for now I think this is easier.

for reference raw type is: 
```
Option<
      ITuple<[PolymeshPrimitivesIdentityClaimClaimType, PolymeshPrimitivesIdentityId]>
 >;
```
and I get `[PolymeshPrimitivesIdentityClaimClaimType, PolymeshPrimitivesIdentityId]` doesn't implement Codec error

File: src/utils/conversion.ts [hunk 19]
```diff
@@ -2870,26 +2873,33 @@ export function transferRestrictionToPolymeshTransferCondition(
   } else if (type === TransferRestrictionType.Percentage) {
     restrictionType = 'MaxInvestorOwnership';
     restrictionValue = percentageToPermill(value as BigNumber, context);
-  } else {
-    const castedValue = value as ClaimCountRestrictionValue;
+  } else if (
+    type === TransferRestrictionType.ClaimCount ||
+    type === TransferRestrictionType.ClaimPercentage
+  ) {
```
Comment (monitz87) [inline src/utils/conversion.ts:2932]: Can't you destructure `min`, `max`, `claim` and `issuer` here at the beginning?

File: src/utils/conversion.ts [hunk 20]
```diff
@@ -3783,17 +3784,18 @@ export function sortTransferRestrictionByClaimValue(
   const getJurisdictionValue = (
     condition: PolymeshPrimitivesTransferComplianceTransferCondition
   ): Option<MeshCountryCode> | undefined => {
-    if (condition.isClaimCount) {
-      if (!condition.asClaimCount[0].isJurisdiction) {
+    const { isClaimCount, asClaimCount, isClaimOwnership, asClaimOwnership } = condition;
```
Comment (monitz87) [inline src/utils/conversion.ts:3855]: You can't destructure the `asX` members right away, they are getters that throw if the `isX` part is false. You have to destructure them inside the corresponding if

File: src/utils/internal.ts [hunk 21]
```diff
@@ -1271,21 +1271,20 @@ export function compareTransferRestrictionToStat(
...
-    }
-  }
 
-  if (type === StatType.ScopedBalance && transferCondition.isClaimOwnership) {
-    const [rawClaim, issuer] = transferCondition.asClaimOwnership;
-    const restrictionIssuerDid = identityIdToString(issuer);
-    const claim = statsClaimToStatClaimUserType(rawClaim);
-    if (restrictionIssuerDid === issuerDid && claim.type === claimType) {
-      return true;
+  if (type === StatType.ScopedCount || type === StatType.ScopedBalance) {
```
Comment (monitz87) [inline src/utils/internal.ts]: What happens if `type === StatType.ScopedCount` but `transferCondition.isClaimOwnership`?

What about

```
let rawClaim, issuer;

if (type === StatType.ScopedCount && transferCondition.isClaimCount) {
  [rawClaim, issuer] = transferCondition.asClaimCount;
}

if (type === StatType.ScopedBalance && transferCondition.isClaimOwnership) {
  [rawClaim, issuer] = transferCondition.asClaimOwnership;
}

if (rawClaim && issuer) {
  // ...
}

return false;
```

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/procedures/addTransferRestriction.ts [hunk 22]
```diff
@@ -90,14 +90,8 @@ export async function prepareAddTransferRestriction(
...
   const maxConditions = u32ToBigNumber(consts.statistics.maxTransferConditionsPerAsset);
-  console.log(
-    'restrictions size',
-    currentRestrictions,
-    currentRestrictions.size,
-    [...currentRestrictions].length
-  );
 
-  const restrictionAmount = new BigNumber(currentRestrictions.size);
+  const restrictionAmount = new BigNumber([...currentRestrictions].length);
```
Comment (monitz87) [inline src/api/procedures/addTransferRestriction.ts:94]: Why this change?

File: src/utils/__tests__/conversion.ts [hunk 23]
```diff
@@ -7852,7 +7851,9 @@ describe('sortByClaimType', () => {
     const countStat = dsMockUtils.createMockStatisticsStatType({ op });
 
     let result = sortStatsByClaimType([jurisdictionStat, accreditedStat, affiliateStat, countStat]);
+    console.log(JSON.stringify(result));
```
Comment (monitz87) [inline src/utils/__tests__/conversion.ts]: This should go

File: src/utils/__tests__/conversion.ts [hunk 24]
```diff
@@ -7852,7 +7851,9 @@ describe('sortByClaimType', () => {
...
     expect(result).toEqual([accreditedStat, affiliateStat, jurisdictionStat, countStat]);
+    // expect(result).toEqual([jurisdictionStat, accreditedStat, affiliateStat, countStat]);
```
Comment (monitz87) [inline src/utils/__tests__/conversion.ts]: Either delete or uncomment

File: src/utils/internal.ts [hunk 25]
```diff
@@ -1257,6 +1257,7 @@ export function compareTransferRestrictionToStat(
   type: StatType,
   claimIssuer?: StatClaimIssuer
 ): boolean {
+  console.log('comparing internally', type, transferCondition);
```
Comment (monitz87) [inline src/utils/internal.ts]: This should go

File: src/api/procedures/__tests__/removeAssetStat.ts [hunk 26]
```diff
@@ -274,6 +283,7 @@ describe('removeAssetStat procedure', () => {
       },
     };
     await expect(prepareRemoveAssetStat.call(proc, args)).rejects.toThrowError(expectedError);
+    console.log('being used done');
```
Comment (monitz87) [inline src/api/procedures/__tests__/removeAssetStat.ts]: This should go

**Review by monitz87 — CHANGES_REQUESTED** [review]

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/types/internal.ts [hunk 27]
```diff
@@ -390,7 +390,7 @@ export enum StatType {
    */
   ScopedCount = 'ScopedCount',
   /**
-   * Keeps track of the amount of supply investors hold who have a certain claim
+   * Keeps track of the percentage of total supply held by investors hold who have a certain claim
```
Comment (monitz87) [inline src/types/internal.ts:393]:
```suggestion
   * Keeps track of the percentage of total supply held by investors who have a certain claim
```

File: src/utils/conversion.ts [hunk 28]
```diff
@@ -2932,21 +2932,22 @@ export function transferRestrictionToPolymeshTransferCondition(
...
-      rawMin = bigNumberToU64(value.min, context);
-      rawMax = optionize(bigNumberToU64)(value.max, context);
+      rawMin = bigNumberToU64(min, context);
+      rawMax = optionize(bigNumberToU64)(max, context);
     } else {
       // i.e. TransferRestrictionType.ClaimPercentage
       restrictionType = 'ClaimOwnership';
-      rawMin = percentageToPermill(value.min, context);
+      rawMin = percentageToPermill(min, context);
       rawMax = percentageToPermill(value.max, context);
```
Comment (monitz87) [inline src/utils/conversion.ts:2944]:
```suggestion
      rawMax = percentageToPermill(max, context);
```

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/utils/conversion.ts (hunk 28 above)
Comment (polymath-eric) [reply src/utils/conversion.ts:2944]: max got destructured earlier, so its type is `BigNumber | undefined` (optional for count).

I did this to satisfy the type checker without needing an assertion, or have another line for the assignment.

Comment (monitz87) [reply src/utils/conversion.ts:2961]: Huh, thought the type checker was smarter. Fair enough

File: src/api/procedures/types.ts [hunk 29]
```diff
@@ -75,6 +79,85 @@ export type GetTransferRestrictionReturnType<T> = ActiveTransferRestrictions<
...
+
+export type AddBalanceStatParams = {
+  type: StatType.Percentage;
+};
+
+export type AddClaimCountStatParams = ClaimCountStatInput & {
+  type: StatType.ScopedCount;
+};
+
+export type AddClaimPercentageStatParams = StatClaimIssuer & {
```
Comment (monitz87) [inline src/api/procedures/types.ts:101]: `AddBalanceStatParams` and `AddClaimPercentageStatParams` have inconsistent naming

File: src/api/procedures/createAsset.ts [hunk 30]
```diff
@@ -224,6 +230,18 @@ export async function prepareCreateAsset(
     );
   }
 
+  if (initialStatistics?.length) {
```
Comment (monitz87) [inline src/api/procedures/createAsset.ts:233]: Even though currently there are no protocol fees for setting active stats, there might be in the future. You should use the `addManualFees` "trick" we're using for all other transactions in this procedure (remember to change the comment to reflect this case)

Comment (monitz87) [reply src/api/procedures/createAsset.ts:233]: Looking back, calling `addManualFees` once per every transaction we add is cumbersome. I think it makes more sense to store all the tags in an array and just make a single call before the `addBatchTransaction` call

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/procedures/createAsset.ts (hunk 30 above)
Comment (polymath-eric) [reply src/api/procedures/createAsset.ts:233]: We are storing the transactions in an array, so with a lookup table, we can map from transactions to tags.

Comment (polymath-eric) [reply src/api/procedures/createAsset.ts:233]: After trying it out, I think we should keep a separate array like you suggested, 

Mapping from `SubmittableExtrinsic` types is vague and harder to reason about. Using `toString` could work, but the extra variable is cleaner imo.

**Review by prashantasdeveloper — APPROVED** [review]

**Review by monitz87 — APPROVED** [review]
